lib/galaxy/web_stack/__init__.py

```diff
@@ -557,6 +569,45 @@ class PasteApplicationStack(ApplicationStack):
     name = 'Python Paste'
 
 
+class GunicornApplicationStack(ApplicationStack):
+    name = "Gunicorn"
+    do_post_fork = "--preload" in os.environ.get("GUNICORN_CMD_ARGS", "") or "--preload" in sys.argv
+    postfork_functions: List[Callable] = []
+    # Will be set to True by external hook
+    late_postfork_event = threading.Event()
+
+    @classmethod
+    def register_postfork_function(cls, f, *args, **kwargs):
+        # do_post_fork determines if we need to run postfork functions
+        if cls.do_post_fork:
+            # if so, we call ApplicationStack.late_postfork once after forking ...
+            if not cls.postfork_functions:
+                os.register_at_fork(after_in_child=cls.late_postfork)
+            # ... and store everything we need to run in ApplicationStack.postfork_functions
+            cls.postfork_functions.append(lambda: f(*args, **kwargs))
+        else:
+            f(*args, **kwargs)
+
+    @classmethod
+    def run_postfork(cls):
+        cls.late_postfork_event.wait(1)
+        for f in cls.postfork_functions:
+            f()
+
+    @classmethod
+    def late_postfork(cls):
+        # We can't run postfork functions immediately, because this is before the gunicorn `post_fork` hook runs,
+        # and we depend on the `post_fork` hook to set a worker id.
+        t = threading.Thread(target=cls.run_postfork)
+        t.start()
+
+    def log_startup(self):
+        msg = [f"Galaxy server instance '{self.config.server_name}' is running"]
+        if "GUNICORN_LISTENERS" in os.environ:
+            msg.append(f'serving on {os.environ["GUNICORN_LISTENERS"]}')
```

Comment:
Testing locally with a default `galaxy.yml`, I only see the first message in the logs.

Reply:
You'll need https://github.com/galaxyproject/gravity/pull/41 for that, or you can run this manually:
```
GALAXY_CONFIG_FILE="config/galaxy.yml"  gunicorn 'galaxy.webapps.galaxy.fast_factory:factory()' --timeout 300 --pythonpath lib -k galaxy.webapps.galaxy.workers.Worker -b localhost:8080 --workers=2 --forwarded-allow-ips='*' --config lib/galaxy/web_stack/gunicorn_config.py --preload
```
(preload is optional, also works without)